File: backend/opencode/catalog.go
package opencode

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/enough/enough/backend/enoughhome"
)

var modelsDevURL = "https://models.dev/api.json"

// minOpencodeGoCatalogModels guards against corrupted/partial cache files (e.g. test mocks).
const minOpencodeGoCatalogModels = 5

const modelsDevProviderGo = "opencode-go"
const modelsDevProviderZen = "opencode"

type modelsDevCatalog struct {
	Providers map[string]modelsDevProvider `json:"-"`
}

type modelsDevProvider struct {
	ID     string                     `json:"id"`
	Models map[string]modelsDevModel `json:"models"`
}

type modelsDevReasoningOption struct {
	Type   string   `json:"type"`
	Values []string `json:"values"`
}

type modelsDevModel struct {
	ID               string                     `json:"id"`
	Name             string                     `json:"name"`
	Family           string                     `json:"family"`
	Reasoning        bool                       `json:"reasoning"`
	ReasoningOptions []modelsDevReasoningOption `json:"reasoning_options"`
	Limit            struct {
		Context int `json:"context"`
		Output  int `json:"output"`
	} `json:"limit"`
	Interleaved      json.RawMessage            `json:"interleaved"`
}

var (
	catalogMu             sync.RWMutex
	opencodeCatalog       map[string]ModelInfo // OpenCode Go (models.dev opencode-go)
	opencodeZenCatalog    map[string]ModelInfo // OpenCode Zen (models.dev opencode)
	catalogLoaded         bool
	backgroundRefreshOnce sync.Once
)

func cacheFilePath() string {
	return filepath.Join(enoughhome.HomeDir(), "cache", "models.json")
}

func loadCatalogFromCache() (goCatalog, zenCatalog map[string]ModelInfo, modTime time.Time, err error) {
	path := cacheFilePath()
	fi, err := os.Stat(path)
	if err != nil {
		return nil, nil, time.Time{}, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, nil, time.Time{}, err
	}
	var all map[string]modelsDevProvider
	if err := json.Unmarshal(data, &all); err != nil {
		return nil, nil, time.Time{}, err
	}
	goCatalog = catalogFromModelsDevProvider(all[modelsDevProviderGo])
	zenCatalog = catalogFromModelsDevProvider(all[modelsDevProviderZen])
	if len(goCatalog) == 0 {
		return nil, nil, time.Time{}, fmt.Errorf("missing go provider in cached catalog")
	}
	return goCatalog, zenCatalog, fi.ModTime(), nil
}

func catalogFromModelsDevProvider(provider modelsDevProvider) map[string]ModelInfo {
	if len(provider.Models) == 0 {
		return nil
	}
	next := make(map[string]ModelInfo, len(provider.Models))
	for id, m := range provider.Models {
		next[id] = modelInfoFromModelsDev(m)
	}
	return next
}

func saveCatalogToCache(data []byte) error {
	path := cacheFilePath()
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}

func startBackgroundRefreshLoop() {
	go func() {
		for {
			time.Sleep(60 * time.Minute)
			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
			_ = RefreshModelsDevCatalog(ctx)
			cancel()
		}
	}()
}

func RefreshModelsDevCatalog(ctx context.Context) error {
	backgroundRefreshOnce.Do(startBackgroundRefreshLoop)

	// 1. Try to load from disk cache first (only if using default URL)
	var cachedGo, cachedZen map[string]ModelInfo
	var modTime time.Time
	var err error
	if modelsDevURL == "https://models.dev/api.json" {
		cachedGo, cachedZen, modTime, err = loadCatalogFromCache()
		if err == nil && len(cachedGo) >= minOpencodeGoCatalogModels {
			// Cache is present. Check freshness (60 minutes).
			if time.Since(modTime) < 60*time.Minute {
				catalogMu.Lock()
				opencodeCatalog = cachedGo
				opencodeZenCatalog = cachedZen
				catalogLoaded = true
				catalogMu.Unlock()
				return nil
			}
		}
	}

	// 2. Fetch from network
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, modelsDevURL, nil)
	if err != nil {
		if cachedGo != nil {
			catalogMu.Lock()
			opencodeCatalog = cachedGo
			opencodeZenCatalog = cachedZen
			catalogLoaded = true
			catalogMu.Unlock()
			return nil
		}
		return err
	}

	client := &http.Client{Timeout: 20 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		if cachedGo != nil {
			catalogMu.Lock()
			opencodeCatalog = cachedGo
			opencodeZenCatalog = cachedZen
			catalogLoaded = true
			catalogMu.Unlock()
			return nil
		}
		return err
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		if cachedGo != nil {
			catalogMu.Lock()
			opencodeCatalog = cachedGo
			opencodeZenCatalog = cachedZen
			catalogLoaded = true
			catalogMu.Unlock()
			return nil
		}
		return err
	}
	if resp.StatusCode >= 400 {
		if cachedGo != nil {
			catalogMu.Lock()
			opencodeCatalog = cachedGo
			opencodeZenCatalog = cachedZen
			catalogLoaded = true
			catalogMu.Unlock()
			return nil
		}
		return fmt.Errorf("models.dev %d: %s", resp.StatusCode, strings.TrimSpace(string(raw)))
	}

	var all map[string]modelsDevProvider
	if err := json.Unmarshal(raw, &all); err != nil {
		if cachedGo != nil {
			catalogMu.Lock()
			opencodeCatalog = cachedGo
			opencodeZenCatalog = cachedZen
			catalogLoaded = true
			catalogMu.Unlock()
			return nil
		}
		return fmt.Errorf("decode models.dev: %w", err)
	}

	goProvider, ok := all[modelsDevProviderGo]
	if !ok {
		if cachedGo != nil {
			catalogMu.Lock()
			opencodeCatalog = cachedGo
			opencodeZenCatalog = cachedZen
			catalogLoaded = true
			catalogMu.Unlock()
			return nil
		}
		return fmt.Errorf("models.dev: missing %q provider", modelsDevProviderGo)
	}

	catalogMu.Lock()
	opencodeCatalog = catalogFromModelsDevProvider(goProvider)
	opencodeZenCatalog = catalogFromModelsDevProvider(all[modelsDevProviderZen])
	catalogLoaded = true
	catalogMu.Unlock()

	if modelsDevURL == "https://models.dev/api.json" {
		_ = saveCatalogToCache(raw)
	}
	return nil
}

func catalogModel(id string) (ModelInfo, bool) {
	return catalogModelForProvider(ProviderOpenCode, id)
}

func catalogModelForProvider(provider, id string) (ModelInfo, bool) {
	catalogMu.RLock()
	defer catalogMu.RUnlock()
	cat := opencodeCatalog
	if provider == ProviderOpenCodeZen {
		cat = opencodeZenCatalog
	}
	if m, ok := cat[id]; ok {
		return m, true
	}
	return ModelInfo{}, false
}

func catalogLoadedOnce() bool {
	catalogMu.RLock()
	defer catalogMu.RUnlock()
	return catalogLoaded
}

func parseReasoningField(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}
	var b bool
	if err := json.Unmarshal(raw, &b); err == nil {
		if b {
			return "reasoning_content"
		}
		return ""
	}
	var obj struct {
		Field string `json:"field"`
	}
	if err := json.Unmarshal(raw, &obj); err == nil {
		return obj.Field
	}
	return ""
}

func modelInfoFromModelsDev(m modelsDevModel) ModelInfo {
	info := ModelInfo{
		ID:             m.ID,
		Name:           m.Name,
		ContextWindow:  m.Limit.Context,
		Reasoning:      m.Reasoning,
		ReasoningField: parseReasoningField(m.Interleaved),
	}
	if info.Name == "" {
		info.Name = titleCaseModelID(m.ID)
	}
	return normalizeModel(info)
}

func opencodeMandatoryThinkingID(id string) bool {
	id = strings.ToLower(id)
	if strings.Contains(id, "deepseek-chat") ||
		strings.Contains(id, "deepseek-reasoner") ||
		strings.Contains(id, "deepseek-r1") ||
		strings.Contains(id, "deepseek-v3") {
		return true
	}
	for _, part := range []string{"minimax", "glm", "kimi", "k2p", "qwen", "big-pickle"} {
		if strings.Contains(id, part) {
			return true
		}
	}
	return false
}



func titleCaseModelID(id string) string {
	parts := strings.Split(id, "-")
	for i, p := range parts {
		if len(p) > 0 {
			parts[i] = strings.ToUpper(p[:1]) + p[1:]
		}
	}
	return strings.Join(parts, " ")
}

File: backend/opencode/catalog.ts
// PORT: mirrors backend/opencode/catalog.go

import { Effect } from "effect";
import fs from "node:fs/promises";
import path from "node:path";
import { home_dir } from "../enoughhome/home";
import { provider_opencode, provider_opencode_zen, type model_info } from "./providers";
import { normalize_model } from "./models";
import { format_thinking_label } from "./thinking";
import type { json_raw_message } from "./types";

export const models_dev_url = "https://models.dev/api.json";
export const min_opencode_go_catalog_models = 5;
export const models_dev_provider_go = "opencode-go";
export const models_dev_provider_zen = "opencode";

export type models_dev_reasoning_option = {
  type: string;
  values: string[];
};

export type models_dev_model = {
  id: string;
  name: string;
  family: string;
  reasoning: boolean;
  reasoning_options?: models_dev_reasoning_option[];
  modalities: {
    input: string[];
  };
  limit: {
    context: number;
    output: number;
  };
  interleaved?: json_raw_message;
};

export type models_dev_provider = {
  id: string;
  models: Record<string, models_dev_model>;
};

let opencode_catalog: Record<string, model_info> = {};
let opencode_zen_catalog: Record<string, model_info> = {};
let catalog_loaded = false;
let background_refresh_started = false;

export const cache_file_path = (): string => {
  return path.join(home_dir(), "cache", "models.json");
};

export const load_catalog_from_cache = (): Effect.Effect<
  { go_catalog: Record<string, model_info>; zen_catalog: Record<string, model_info>; mod_time: Date },
  Error
> => {
  return Effect.tryPromise({
    try: async () => {
      const p = cache_file_path();
      const stat = await fs.stat(p);
      const data = await fs.readFile(p, "utf8");
      const all = JSON.parse(data) as Record<string, models_dev_provider>;

      const go_catalog = catalog_from_models_dev_provider(all[models_dev_provider_go]);
      const zen_catalog = catalog_from_models_dev_provider(all[models_dev_provider_zen]);

      if (Object.keys(go_catalog).length === 0) {
        throw new Error("missing go provider in cached catalog");
      }
      return { go_catalog, zen_catalog, mod_time: stat.mtime };
    },
    catch: (cause) => cause instanceof Error ? cause : new Error(String(cause)),
  });
};

export const catalog_from_models_dev_provider = (provider?: models_dev_provider): Record<string, model_info> => {
  if (!provider || !provider.models || Object.keys(provider.models).length === 0) {
    return {};
  }
  const next: Record<string, model_info> = {};
  for (const [id, m] of Object.entries(provider.models)) {
    next[id] = model_info_from_models_dev(m);
  }
  return next;
};

export const save_catalog_to_cache = (data: string): Effect.Effect<void, Error> => {
  return Effect.tryPromise({
    try: async () => {
      const p = cache_file_path();
      const dir = path.dirname(p);
      await fs.mkdir(dir, { recursive: true });
      await fs.writeFile(p, data, "utf8");
    },
    catch: (cause) => cause instanceof Error ? cause : new Error(String(cause)),
  });
};

const start_background_refresh_loop = () => {
  setInterval(async () => {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 30000);
    try {
      await Effect.runPromise(refresh_models_dev_catalog(controller.signal));
    } catch {} finally {
      clearTimeout(timeout);
    }
  }, 60 * 60 * 1000); // 60 minutes
};

export const refresh_models_dev_catalog = (ctx?: AbortSignal): Effect.Effect<void, Error> => {
  if (!background_refresh_started) {
    background_refresh_started = true;
    start_background_refresh_loop();
  }

  return Effect.tryPromise({
    try: async () => {
      // 1. Try to load from disk cache first (only if using default URL)
      let cached_go: Record<string, model_info> | null = null;
      let cached_zen: Record<string, model_info> | null = null;
      let mod_time: Date | null = null;

      if (models_dev_url === "https://models.dev/api.json") {
        try {
          const cache_res = await Effect.runPromise(load_catalog_from_cache());
          cached_go = cache_res.go_catalog;
          cached_zen = cache_res.zen_catalog;
          mod_time = cache_res.mod_time;

          if (Object.keys(cached_go).length >= min_opencode_go_catalog_models) {
            const age_ms = Date.now() - mod_time.getTime();
            if (age_ms < 60 * 60 * 1000) { // 60 minutes
              opencode_catalog = cached_go;
              opencode_zen_catalog = cached_zen;
              catalog_loaded = true;
              return;
            }
          }
        } catch {}
      }

      // 2. Fetch from network
      try {
        const resp = await fetch(models_dev_url, { signal: ctx });
        const raw = await resp.text();
        if (resp.status >= 400) {
          if (cached_go !== null && cached_zen !== null) {
            opencode_catalog = cached_go;
            opencode_zen_catalog = cached_zen;
            catalog_loaded = true;
            return;
          }
          throw new Error(`models.dev ${resp.status}: ${raw.trim()}`);
        }
        const all = JSON.parse(raw) as Record<string, models_dev_provider>;
        const go_provider = all[models_dev_provider_go];
        if (!go_provider) {
          if (cached_go !== null && cached_zen !== null) {
            opencode_catalog = cached_go;
            opencode_zen_catalog = cached_zen;
            catalog_loaded = true;
            return;
          }
          throw new Error(`models.dev: missing "${models_dev_provider_go}" provider`);
        }

        opencode_catalog = catalog_from_models_dev_provider(go_provider);
        opencode_zen_catalog = catalog_from_models_dev_provider(all[models_dev_provider_zen]);
        catalog_loaded = true;

        if (models_dev_url === "https://models.dev/api.json") {
          try {
            await Effect.runPromise(save_catalog_to_cache(raw));
          } catch {}
        }
      } catch (err) {
        if (cached_go !== null && cached_zen !== null) {
          opencode_catalog = cached_go;
          opencode_zen_catalog = cached_zen;
          catalog_loaded = true;
          return;
        }
        throw err;
      }
    },
    catch: (cause) => cause instanceof Error ? cause : new Error(String(cause)),
  });
};

export const catalog_model = (id: string): [model_info, boolean] => {
  return catalog_model_for_provider(provider_opencode, id);
};

export const catalog_model_for_provider = (provider: string, id: string): [model_info, boolean] => {
  let cat = opencode_catalog;
  if (provider === provider_opencode_zen) {
    cat = opencode_zen_catalog;
  }
  const m = cat[id];
  if (m !== undefined) {
    return [m, true];
  }
  return [{ id: "", name: "", context_window: 0, reasoning: false }, false];
};

export const catalog_loaded_once = (): boolean => {
  return catalog_loaded;
};

export const parse_reasoning_field = (raw?: json_raw_message | unknown): string => {
  if (raw === undefined || raw === null) {
    return "";
  }
  // Already parsed from models.dev JSON (interleaved object/boolean).
  if (typeof raw === "boolean") {
    return raw ? "reasoning_content" : "";
  }
  if (typeof raw === "object" && !Array.isArray(raw) && !(raw instanceof Uint8Array)) {
    const val = raw as Record<string, unknown>;
    if ("field" in val && typeof val.field === "string") {
      return val.field;
    }
    return "";
  }
  if (!(raw instanceof Uint8Array) && !Buffer.isBuffer(raw)) {
    return "";
  }
  const buf = raw instanceof Uint8Array ? raw : new Uint8Array(raw as Buffer);
  if (buf.length === 0) {
    return "";
  }
  const text = new TextDecoder().decode(buf);
  try {
    const val = JSON.parse(text);
    if (typeof val === "boolean") {
      return val ? "reasoning_content" : "";
    }
    if (val && typeof val === "object" && "field" in val && typeof val.field === "string") {
      return val.field;
    }
  } catch {}
  return "";
};

export const model_info_from_models_dev = (m: models_dev_model): model_info => {
  const info: model_info = {
    id: m.id,
    name: m.name,
    context_window: m.limit.context,
    reasoning: m.reasoning,
    reasoning_field: parse_reasoning_field(m.interleaved),
    supports_images: model_supports_images_from_modalities(m.modalities.input),
  };
  if (info.name === "") {
    info.name = title_case_model_id(m.id);
  }
  if (m.reasoning_options !== undefined && m.reasoning_options.length > 0) {
    for (const option of m.reasoning_options) {
      if (option.values !== undefined && option.values.length > 0) {
        info.thinking_levels = option.values.map((v) => ({
          id: v,
          name: format_thinking_label(v as any),
        }));
        info.reasoning = true;
        break;
      }
    }
  }
  return normalize_model(info);
};

export const model_supports_images_from_modalities = (input: string[]): boolean => {
  return input.includes("image");
};

export const title_case_model_id = (id: string): string => {
  return id
    .split("-")
    .map((p) => (p.length > 0 ? p[0].toUpperCase() + p.slice(1) : p))
    .join(" ");
};

export const get_catalog_models = (provider: string): model_info[] => {
  const cat = provider === provider_opencode_zen ? opencode_zen_catalog : opencode_catalog;
  return Object.values(cat);
};

/*
PORT STATUS
source path: backend/opencode/catalog.go
source lines: 328
draft lines: 231
confidence: high
status: phase_b_compile
*/
